apis/crd/v1alpha1/cloudentityselector_webhook.go

```diff
@@ -218,28 +218,123 @@ func (r *CloudEntitySelector) validateMatchSections() error {
-					for _, vmMatch := range m.VMMatch {
-						if len(strings.TrimSpace(vmMatch.MatchID)) != 0 ||
-							len(strings.TrimSpace(vmMatch.MatchName)) != 0 {
-							return fmt.Errorf("vpc matchName with either vm matchID" +
-								" or vm matchName is not supported, use vpc matchID instead" +
-								" of vpc matchName")
-						}
+			if m.VpcMatch != nil && len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
+				for _, vmMatch := range m.VMMatch {
+					if len(strings.TrimSpace(vmMatch.MatchID)) != 0 ||
+						len(strings.TrimSpace(vmMatch.MatchName)) != 0 {
+						return fmt.Errorf("vpc matchName with either vm matchID" +
+							" or vm matchName is not supported, use vpc matchID instead" +
+							" of vpc matchName")
 					}
 				}
 			}
 		}
 	}
+
+	// Block ambiguous match combinations as it can result in unpredictable behavior wrt "agented" configuration.
+	// vpcIDOnlyMatch map holds VPC ID as key if the selector contains only vpcMatch matchID.
+	// vmIDOnlyMatch map holds VM ID as key if the selector contains only vmMatch matchID.
+	// vmNameOnlyMatch map holds VM Name as key if the selector contains only vmMatch matchName.
+	// vmIDWithVpcMatch map holds VM ID and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchID.
+	// vmNameWithVpcMatch map holds VM Name and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchName.
+	vpcIDOnlyMatch := make(map[string]struct{})
+	vmIDOnlyMatch := make(map[string]struct{})
+	vmNameOnlyMatch := make(map[string]struct{})
+	vmIDWithVpcMatch := make(map[string]struct{})
+	vmNameWithVpcMatch := make(map[string]struct{})
+
+	for _, m := range r.Spec.VMSelector {
+		if m.VpcMatch != nil {
+			if err := r.validateVPCMatch(m, vmIDOnlyMatch, vpcIDOnlyMatch, vmIDWithVpcMatch,
+				vmNameWithVpcMatch); err != nil {
+				return err
+			}
+		} else {
+			if err := r.validateVMMatch(m, vmIDOnlyMatch, vmNameOnlyMatch, vmIDWithVpcMatch); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// validateVPCMatch function validates selector with VpcMatch section to find if it conflicts with existing vmSelectors.
+func (r *CloudEntitySelector) validateVPCMatch(selector VirtualMachineSelector, vmIDOnlyMatch map[string]struct{},
+	vpcIDOnlyMatch map[string]struct{}, vmIDWithVpcMatch map[string]struct{},
+	vmNameWithVpcMatch map[string]struct{}) error {
+	// Block same VPC ID configuration in two selectors with only vpcMatch section.
+	// Block same VM ID configuration in any two selectors, same is not applicable for VM Name as VM Name need not be unique.
+	// Block same combination of VPC ID and VM ID configuration in any two selectors.
+	// Block same combination of VPC ID and VM Name configuration in any two selectors.
+	if selector.VpcMatch.MatchID != "" {
+		if len(selector.VMMatch) == 0 {
+			// Make sure there is no selector with same VPC ID with only vpcMatch section.
+			if _, found := vpcIDOnlyMatch[selector.VpcMatch.MatchID]; found {
```

Comment:
No need add else block, since if its not found you want to add it to the map.
```suggestion
	
	if _, found := vpcIDOnlyMatch[selector.VpcMatch.MatchID]; found {
		return fmt.Errorf("same vpcMatch matchID %v configured in two match selectors",
			selector.VpcMatch.MatchID)
	}
	vpcIDOnlyMatch[selector.VpcMatch.MatchID] = struct{}{}
```

Reply:
err is already used before, hence had to change the variable name.

Reply:
agree